src/main/java/com/feedzai/commons/sql/abstraction/batch/AbstractBatch.java

```diff
@@ -199,10 +199,32 @@ public void add(final String entityName, final EntityEntry ee) throws DatabaseEn
 
     /**
      * Flushes the pending batches.
+     *
+     * @implSpec Same as {@link #flush(boolean)} with {@link false}.
      */
     public void flush() {
+        flush(false);
+    }
+
+    /**
+     * Flushes the pending messages.
+     * <p>
+     * If {@param sync} is {@literal true} it waits for other pending flush operations.
+     * <p>
+     * If {@param sync} is {@literal false} it can return directly if the buffer if the batch is empty.
+     *
+     * @param sync {@literal true} if it should wait for other {@link #flush()} operations.
+     * @implSpec It is possible that two threads might execute in competing order and sync execution acquires the flushTransactionLock before a non synchronous one leading
+     * to a non serial execution.
+     * @since 2.1.6
+     */
+    public void flush(boolean sync) {
         List<BatchEntry> temp;
 
+        if (sync) {
+            flushTransactionLock.lock();
```

Comment:
Ideally this lock should be followed by a try finally, in case something goes wrong.